File: source/2PlayersTicTacToe/2PlayersTicTacToe.hs
import Data.List

{- Definizioni di nuovi tipi -}


{- Box rappresenta una casella della tabella: può essere vuota o occupata da un giocatore (X o 0).
Una casella vuota è rappresentata dal placeholder (-) -}

data Box = Empty | Full Player
    deriving Eq

{- Player rappresenta il simbolo di un giocatore -}

data Player = X | O
    deriving (Show, Eq)

{- Per visualizzare le mosse sulla tabella creiamo un'istanza della typeclass Show per il tipo Box 
gli spazi ai lati sono aggiunti per centrare i simboli nelle rispettive caselle -}

instance Show Box where
    show Empty =    "  -  "
    show (Full X) = "  X  "
    show (Full O) = "  O  " 


{- Definizioni di funzioni-}


{- Tabella vuota utilizzata all'inizio di una partita -}

emptyBoard :: [Box]
emptyBoard = [Empty, Empty, Empty,
              Empty, Empty, Empty,
              Empty, Empty, Empty]


{- Ritorna una riga della tabella sotto forma di stringa: le caselle sono separate dal carattere "|" -}

renderRow :: [Box] -> String
renderRow row = intercalate "|" $ map show row


{- Stringa separatrice tra righe -}

dividingLine :: String
dividingLine =  "   _____|_____|_____\n" ++ "        |     |     "


{-Indici di colonna -}

topCoordsLine :: String
topCoordsLine = "     1     2     3\n"


{- Funzione per visualizzare la tabella -}

renderBoard :: [Box] -> IO ()
renderBoard board = do
  putStrLn topCoordsLine  
  putStrLn $ "A  " ++ renderRow firstRow
  putStrLn dividingLine
  putStrLn $ "B  " ++ renderRow secondRow
  putStrLn dividingLine
  putStrLn $ "C  " ++ renderRow thirdRow
  where firstRow  = take 3 board
        secondRow = drop 3.take 6 $ board
        thirdRow  = drop 6 board


{- Lista di mosse accettabili -}

validMoves :: [String]
validMoves = ["A1", "A2", "A3", "B1", "B2", "B3", "C1", "C2", "C3"]


{- Funzione che mappa una mossa valida al corrispondente indice di casella, ritorna -1 per mosse invalide -}

getBoxIndex :: String -> Int
getBoxIndex "A1" =  0
getBoxIndex "A2" =  1
getBoxIndex "A3" =  2
getBoxIndex "B1" =  3
getBoxIndex "B2" =  4
getBoxIndex "B3" =  5
getBoxIndex "C1" =  6
getBoxIndex "C2" =  7
getBoxIndex "C3" =  8
getBoxIndex _ = -1


{- Funzione che valida una mossa controllando se appartiene alla lista validMoves -}

validateMove :: String -> Bool
validateMove move = move `elem` validMoves


{- Funzione che controlla se una casella è occupata -}

isEmpty :: Box -> Bool
isEmpty Empty = True
isEmpty (Full _) = False


{- Funzione che data una tabella e un indice, restituisce True se la casella a quell'indice è libera -}

validatePosition :: [Box] -> String -> Bool

validatePosition board m | i == -1 = False
                         | isEmpty $ board !! i = True
                         | otherwise = False
    where i = getBoxIndex m  
 

{- Funzione che prende una lista e un indice, e restituisce una coppia formata dalle liste degli elementi a sinistra e a destra di quello all'indice specificato -}

splitAtIndex :: [a] -> Int -> ([a], [a])

splitAtIndex lst i = (left, right)
    where 
        (left, xs) = splitAt i lst
        right = drop 1 xs 


{- Funzione per l'inserimento di un simbolo Player nella tabella di gioco -}

playMove :: [Box] -> Box -> Int -> [Box] 

playMove board player i = xs ++ [player] ++ ys
    where (xs, ys) = splitAtIndex board i     


{- Funzione che prende una mossa dall'utente e restituisce l'indice di tabella corrispondente -}

getMove :: [Box] -> Player -> IO Int

getMove board currentPlayer = do
    putStrLn $ "\nPlayer " ++ (show currentPlayer) ++ " enter a move (A1 - C3): "
    move <- getLine
    if validateMove move && validatePosition board move 
        then return $ getBoxIndex move
    else do
        putStrLn "\nInvalid move, try again!"
        getMove board currentPlayer


{- Funzione che gestisce l'alternanza dei turni -}

nextPlayer :: Player -> Player

nextPlayer X = O
nextPlayer O = X


{- Funzione che controlla se nella tabella attuale è presente una combinazione vincente -}

checkWinner :: Player -> [Box] -> Bool

checkWinner move board =
  or 
    [board !! 0 == (Full move) && board !! 1 == (Full move) && board !! 2 == (Full move),
     board !! 3 == (Full move) && board !! 4 == (Full move) && board !! 5 == (Full move),
     board !! 6 == (Full move) && board !! 7 == (Full move) && board !! 8 == (Full move),
     board !! 0 == (Full move) && board !! 3 == (Full move) && board !! 6 == (Full move),
     board !! 1 == (Full move) && board !! 4 == (Full move) && board !! 7 == (Full move),
     board !! 2 == (Full move) && board !! 5 == (Full move) && board !! 8 == (Full move),
     board !! 0 == (Full move) && board !! 4 == (Full move) && board !! 8 == (Full move),
     board !! 6 == (Full move) && board !! 4 == (Full move) && board !! 2 == (Full move)]


{- Funzione che controlla se la partita è un pareggio -}

checkTie :: [Box] -> Bool 

checkTie board = all (\box -> not (isEmpty box)) board 


{- Funzione che comunica all'utente l'evento di un pareggio o di una vittoria, in caso contrario chiama ricorsivamente gameLoop con il giocatore successivo -}

checkGameState :: [Box] -> Player -> IO ()

checkGameState board currentPlayer | checkTie board      = putStrLn "\nIt's a tie!"
                                   | checkWinner X board = putStrLn "\nPlayer X won!" 
                                   | checkWinner O board = putStrLn "\nPlayer O won!"
                                   | otherwise = gameLoop board (nextPlayer currentPlayer)


{- Funzione principale che gestice lo stato della partita-}

gameLoop :: [Box] -> Player -> IO ()
gameLoop board currentPlayer = do
    choice <- getMove board currentPlayer
    {- Creiamo una nuova tabella con il simbolo inserito dal giocatore -}
    let newBoard = playMove board (Full currentPlayer) choice 
    renderBoard newBoard
    {- Controlliamo un'eventuale vittoria o pareggio, se nessuno di questi eventi si verifica sarà richiamato ricorsivamente gameLoop con il giocatore successivo -}
    checkGameState newBoard currentPlayer


{- Funzione che chiede all'utente se vuole fare un'altra partita, se l'utente accetta è richiamata main -}

newGame :: IO () 
newGame = do
    putStrLn "\nPlay again? (y/n):"
    choice <- getLine
    if choice == ['y'] then main
    else if choice == ['n'] then return ()
    else do
        putStrLn "\nInvalid character, try again!"
        newGame


main :: IO ()
main = do
    putStrLn "\nWelcome to tic tac toe, good luck and have fun!\n"
    renderBoard emptyBoard
    gameLoop emptyBoard X 
    {- Alla fine di una partita chiediamo all'utente se vuole giocare di nuovo -}
    newGame
    putStrLn "\nBye Bye!" 
    

File: source/2PlayersTicTacToe/prolog/2PlayersTicTacToe.pl
/* predicati per rappresentare lo stato delle caselle */

% una casella e' occupata da un giocatore se contiene una x oppure una o

x(B):- B = x.
o(B):- B = o.
player(B) :- x(B); o(B).

% una casella e' piena se e' occupata da un giocatore, in caso contrario e' vuota

full(B)  :- player(B).
empty(B) :- \+(full(B)).

/* predicati per la visualizzazione dello stato della tabella */

% visualizza lo stato delle caselle

showBox(B) :- o(B), write('  O  ').
showBox(B) :- x(B), write('  X  ').
showBox(B) :- empty(B), write('  -  ').

% visualizza lo stato delle righe
% X, Y, Z: caselle che costituiscono una riga

showRow(X, Y, Z) :-
    showBox(X),
    write('|'),
    showBox(Y),
    write('|'),
    showBox(Z), nl.

% visualizza la riga separatrice 

showDividingLine :-
    write('   _____|_____|_____'), nl,
    write('        |     |     '), nl.

% visualizza gli indici di colonna

showColIndex :-
    nl,write('     1     2     3'), nl, nl.

% visualizza lo stato della tabella corrente
% la lista in input rappresenta la tabella di gioco

showBoard([A, B, C, D, E, F, G, H, I]) :-
    showColIndex,
    write('A  '),
    showRow(A, B, C),
    showDividingLine,
    write('B  '),
    showRow(D, E, F),
    showDividingLine,
    write('C  '),
    showRow(G, H, I).

/* controlli sull'input */

% controlla se la casella all'indice specificato e' vuota
% I: indice di lista (1 - 9) 
% B: lista che rappresenta la tabella di gioco corrente

isEmpty(I, B) :- 
    nth(I, B, E), % successo se l'I-esimo elemento di B e' uguale ad E
    empty(E).
    
% ottiene l'indice corrispondente ad una mossa inserita dall'utente

getBoxIndex('a1', 1).
getBoxIndex('a2', 2).
getBoxIndex('a3', 3).
getBoxIndex('b1', 4).
getBoxIndex('b2', 5).
getBoxIndex('b3', 6).
getBoxIndex('c1', 7).
getBoxIndex('c2', 8).
getBoxIndex('c3', 9).

/* logica di gioco */

% sostituisce l'N-esimo elemento di una lista con un nuovo elemento
% argomenti: indice, elemento da inserire, lista corrente, nuova lista
% caso base: sostitizione del primo elemento
% caso ricorsivo: se l'indice e' maggiore di 1, si richiama il predicato sulle code delle liste con I diminuito di 1, fino a ricondursi al caso base

replaceAtIndex(1, E, [_ | T], [E | T]).
replaceAtIndex(I, E, [H | T], [H | R]) :-
    I > 1, 
    INew is I - 1, 
    replaceAtIndex(INew, E, T, R).

% legge una mossa dall'utente e determina l'indice corrispondente
% B: lista che rappresenta la tabella di gioco corrente
% P: giocatore attivo nel turno corrente (x oppure o)
% I: indice corrispondente alla mossa letta

getMove(B, P, I) :- 
    player(P),
    ((x(P),             % turno del giocatore X
    nl, write('Player X enter a move (a1 - c3): '));
    (o(P),              % turno del giocatore O
    nl, write('Player O enter a move (a1 - c3): '))),
    read(M),    
    getBoxIndex(M, I1), % ottiene il corrispondente indice di lista
    isEmpty(I1, B),     % controlla che la casella corrispondente sia vuota
    I is I1.
getMove(B, P, I) :-
    player(P),
    nl, write('Invalid move, try again!'), nl,
    getMove(B, P, I).

% gestisce l'alternanza dei giocatori

nextPlayer(x, o).
nextPlayer(o, x).

% controlla se e' stata raggiunta una condizione di vittoria

checkWin(B, P) :- 
    rowWin(B, P);
    colWin(B, P);
    diagWin(B, P).

% determina le condizioni di vittoria
% B: tabella corrente
% P: simbolo di un giocatore

% 3 simboli uguali allineati in riga
rowWin(B, P) :-
    player(P), 
    (B = [P, P, P, _, _, _, _, _, _];
     B = [_, _, _, P, P, P, _, _, _];
     B = [_, _, _, _, _, _, P, P, P]).

% 3 simboli uguali allineati in colonna
colWin(B, P) :-
    player(P), 
    (B = [P, _, _, P, _, _, P, _, _];
     B = [_, P, _, _, P, _, _, P, _];
     B = [_, _, P, _, _, P, _, _, P]).

% 3 simoli uguali allineati in diagonale 
diagWin(B, P) :-
    player(P), 
    (B = [P, _, _, _, P, _, _, _, P];
     B = [_, P, _, _, P, _, _, P, _];
     B = [_, _, P, _, P, _, P, _, _]).

% controlla se e' stata raggiunta la condizione di parita' (tutte le caselle occupate da un giocatore)

checkTie(B) :- maplist(player, B).

% controlla lo stato del gioco ad ogni turno comunicando all'utente un'eventuale vittoria o pareggio
% se non e' riscontrata una condizione di fine partita, viene chiamato ricorsivamente gameLoop per iniziare un nuovo turno
% B: tabella corrente
% P: giocatore corrente

% controlla vittoria del giocatore x
checkGameState(B, P) :-
    checkWin(B, P),
    x(P),
    nl, write('Player X won!').
% controlla vittoria del giocatore o
checkGameState(B, P) :-
    checkWin(B, P),
    o(P),
    nl, write('Player O won!').
% controlla parita'
checkGameState(B, _) :-
    checkTie(B),
    nl, write('It\'s a tie!').
% se tutte le precedenti falliscono, continua con il prossimo turno
checkGameState(B, P) :-
    nextPlayer(P, NP),
    gameLoop(B, NP).

% gestisce lo stato della partita
% B: tabella corrente
% P: giocatore corrente

gameLoop(B, P) :-
    getMove(B, P, I),            % acquisice una mossa da giocatore corrente
    replaceAtIndex(I, P, B, NB), % genera una nuova lista con la mossa inserita dal giocatore
    showBoard(NB),               % mostra la nuova tabella
    checkGameState(NB, P).       % controlla il raggiungimento di una condizione di fine partita

% chiede all'utente se vuole giocare ancora una volta finita una partita
% valutato una volta raggiunta una condizione di fine partita (parita' oppure vittoria)

newGame :-
    nl, write('Play again? (y/n): '),
    read(C),
    ((C == 'y', % se il carattere letto e' y, chiama ricorsivamente la funzione main
     main);
    (C == 'n',  % altrimenti, se il carattere letto e' n, termina
     nl, write('Bye Bye!'))).
newGame :-
    nl, write('Invalid character, try again!'),
    newGame.

% predicato main, da eseguire per lanciare il programma

main :- 
    nl, write('Welcome to tic tac toe, good luck and have fun!'), nl,
    write('Warning: input must be lowercase and end with a period'), nl,
    showBoard([e,e,e,e,e,e,e,e,e]),   % stampa la tabella vuota
    gameLoop([e,e,e,e,e,e,e,e,e], x), % inizia il gioco con la tabella vuota e il giocatore x
    newGame.